### DISCUSSION TITLE
Can we use the terminology "FIPS mode"?

### DISCUSSION THREAD

**Discussion by junaruga** [openssl/openssl#21797] [Q&A] [answered]
I might hear that the terminology "FIPS mode" is Red Hat specific in somewhere in the past.

However, when I searched this repository by the "fips mode", I can see the the terminology is used in many parts. Below is the result on the latest master branch c2a8226cba2757b251729620aedffeed23d73623. My question is which context we can use "FIPS mode" in, and which context we should use "FIPS (case or enabled)" instead of "FIPS mode" in? 

```
$ git remote -v | grep origin
origin	https://github.com/openssl/openssl.git (fetch)
origin	https://github.com/openssl/openssl.git (push)

$ grep -rin 'fips mode' 2>/dev/null | wc -l
92

$ grep -rin 'fips mode' 2>/dev/null
providers/implementations/keymgmt/dh_kmgmt.c:67:    int generator; /* Used by DH_PARAMGEN_TYPE_GENERATOR in non fips mode only */
providers/implementations/keymgmt/dh_kmgmt.c:751:             * NOTE: The old safe prime generator code is not used in fips mode,
providers/implementations/keymgmt/rsa_kmgmt.c:245:/* In fips mode there are no multi-primes. */
providers/implementations/kdfs/pbkdf2_fips.c:14: * Extra checks are done by default in fips mode only.
...
```

The reason is because I am working for [OpenSSL Ruby binding](https://github.com/ruby/openssl) using this terminology in some parts in the repository. And I want to use the terminology properly in the repository. Below is the result on the OpenSSL Ruby binding (ruby/openssl) latest master branch `bff06067cd3cf611b50cb76adc2ae40f9406cb1e`.

```
$ git remote -v | grep origin
origin	git@github.com:ruby/openssl.git (fetch)
origin	git@github.com:ruby/openssl.git (push)

$ grep -rin 'fips mode' 2>/dev/null ext/ test/ .github/
ext/openssl/ossl.c:427: * Turns FIPS mode on or off. Turning on FIPS mode will obviously only have an
ext/openssl/ossl.c:432: *   OpenSSL.fips_mode = true   # turn FIPS mode on
ext/openssl/ossl.c:441:            ossl_raise(eOSSLError, "Turning on FIPS mode failed");
ext/openssl/ossl.c:445:            ossl_raise(eOSSLError, "Turning off FIPS mode failed");
ext/openssl/ossl.c:453:	    ossl_raise(eOSSLError, "Turning on FIPS mode failed");
ext/openssl/ossl.c:456:	    ossl_raise(eOSSLError, "Turning off FIPS mode failed");
ext/openssl/ossl.c:461:	ossl_raise(eOSSLError, "This version of OpenSSL does not support FIPS mode");
test/openssl/test_fips.rb:9:      omit "Only for FIPS mode environment"
test/openssl/test_fips.rb:13:      assert OpenSSL.fips_mode == true, ".fips_mode should return true on FIPS mode enabled"
test/openssl/test_fips.rb:19:      omit "Only for non-FIPS mode environment"
test/openssl/test_fips.rb:23:      message = ".fips_mode should return false on FIPS mode disabled. " \
test/openssl/test_fips.rb:24:                "If you run the test on FIPS mode, please set " \
test/openssl/test_fips.rb:46:        pend "Could not set FIPS mode (OpenSSL::OpenSSLError: \#$!); skipping"
.github/workflows/test.yml:179:      # Run only the passing tests on the FIPS mode as a temporary workaround.
.github/workflows/test.yml:180:      # TODO Fix other tests, and run all the tests on FIPS mode.
.github/workflows/test.yml:181:      - name: test on fips mode
```

**Comment by t8m** [comment, marked as answer]
In general the `FIPS mode` term should not be used in the OpenSSL code. Of course, it still is, but if anybody would like to submit a PR replacing it, it would be welcome. In most cases the replacement would be simple - `FIPS mode` -> `FIPS module` or `FIPS provider` - these two are equivalent and interchangeable.

There is no such thing as FIPS mode anymore - in the upstream OpenSSL that is. A FIPS mode is now application specific or os vendor specific thing.

**Reply by junaruga** [reply]
> There is no such thing as FIPS mode anymore - in the upstream OpenSSL that is. A FIPS mode is now application specific or os vendor specific thing.

Thanks for your answer. The "anymore" means the upstream OpenSSL was using the "FIPS mode" intentionally right? The last item used the "FIPS mode" is in the changes between 1.0.2h and 1.1.0.

https://github.com/openssl/openssl/blob/617cab094f0f0d4e71f8b9da5663be8ab06cba92/CHANGES.md?plain=1#L5341-L5342

```
 4227 ### Changes between 1.0.2h and 1.1.0  [25 Aug 2016]
 4228
...
 5341  * Add selftest checks and algorithm block of non-fips algorithms in
 5342    FIPS mode. Remove DES2 from selftests.
```

In my assumption, in OpenSSL 1.0, we needed to enable FIPS only in configuration step of the installation, and we couldn't enable/disable FIPS after the installation. And we called it as "FIPS mode". Now we don't use the terminology any more, because we can enable/disable FIPS just by loading or not loading FIPS module dynamically by the OpenSSL configuration file. My assumption is correct?

**Reply by mattcaswell** [reply]
In 1.0.x you were able to use the "old" FIPS module (i.e. the 2.0.x FIPS module). In those versions the concept of FIPS mode existed. You were either in FIPS mode, in which case FIPS algorithms were always used, or you were not in FIPS mode and we used the normal default algorithms. There were API functions `FIPS_mode()` and `FIPS_mode_set()`.

In OpenSSL 1.1.0 there was no FIPS module at all (so that CHANGES entry is misleading). So the "FIPS mode" was always off.

In OpenSSL 3.0 the FIPS support was rewritten from scratch. In 3.x there is no concept of FIPS mode because applications can have the flexibility to use FIPS algorithms in parts of the application, and default algorithms in other parts at the same time i.e. its not a universal setting, so FIPS mode no longer makes sense from an application perspective.

That said you will still find lots of references to "fips mode" internally to the OpenSSL codebase. Although it doesn't make sense from an application perspective it kind of does internally. Some bits of code get compiled *twice* (with different pre-processor macros): once for inclusion in the default provider and once for inclusion in the FIPS provider. Due to the different pre-processor settings the different compilations will work subtly differently to each other. It is common to find comments referring to "fips mode", where it is used informally to say "when compiled for the FIPS provider we do x, when compiled for the default provider we do y".

**Reply by t8m** [reply]
> It is common to find comments referring to "fips mode", where it is used informally to say "when compiled for the FIPS provider we do x, when compiled for the default provider we do y".

These cases should be rewritten to use fips module or fips provider to be less misleading.

**Comment by t8m** [comment]
Maybe to make it more clear - an application or OS can still have a FIPS mode and be using OpenSSL 3.x. However it is the matter of the application or the OS how it ensures that only the algorithms from the OpenSSL 3.x FIPS provider are used in a FIPS compliant manner (according to the FIPS security policy of the OpenSSL FIPS module).

**Reply by junaruga** [reply]
I can understand that. For example, in Red Hat Enterprise Linux (RHEL) downstream OpenSSL, there is a switch `OPENSSL_FORCE_FIPS_MODE=1` to load FIPS provider and another provider to run OpenSSL FIPS. In this case, an application (downstream OpenSSL) is using this as "FIPS mode". And RHEL has a feature to enable/disable FIPS in OS level. And it is called "FIPS mode".

**Reply by nhorman** [reply]
I believe your statement is fairly accurate.  "FIPS mode" in Red Hat terminology was the process by which you informed a system on boot that you wanted to comply with the FIPS certification claims that Red Hat made when they certified their OS.  It did many things, including, but not limited to setting up Openssl to only use FIPS approved algorithms.  It also ensured that the kernel only loaded FIPS approved crypto algs, did an integrity check on the rootfs, etc, which matched the claims they made in their compliance documents (which I believe you can read on the NIST website).  

And, just to say it aloud (though I'm guessing you're clear on this already), it should also be noted that using the FIPS provider in openssl doesn't automatically make your application FIPS compliant or certified, it just means that you are telling openssl to only use FIPS approved ciphers/hashes/keys, etc.  Certification is still the responsibility of the application author.
